### PR TITLE
Parse server error responses

### PR DESCRIPTION
**Pull request by sitaram-kalluri**
**- What I did**
Parse JSON server responses in at_lookup

### CODE REVIEW

File: at_lookup/lib/src/at_lookup_impl.dart [hunk 1]
```diff
@@ -267,6 +267,19 @@ class AtLookupImpl implements AtLookUp {
     }
     // If response starts with error:, throw AtLookupException.
     if (_isError(verbResult)) {
+      // If server returns the error response in JSON Encoded string, decodes the JSON String
+      // and throws the appropriate AtException basing on the error code.
+      if (verbResult.contains('errorCode')) {
```
Comment (murali-shris) [inline at_lookup/lib/src/at_lookup_impl.dart]: contains check may fail if some random data from server has errorCode in it..
error: < text > - old format
error: < json > - new format
_isError(verbResult)) can have the previous implementation..if json parsing fails, then consider error response to be in old format

**Review by VJag — APPROVED** [review]

File: at_lookup/lib/src/at_lookup_impl.dart (hunk 1 above)
Comment (sitaram-kalluri) [reply at_lookup/lib/src/at_lookup_impl.dart]: Modified the code to have error: before JSON response.

File: at_lookup/lib/src/at_lookup_impl.dart [hunk 2]
```diff
@@ -267,19 +267,31 @@ class AtLookupImpl implements AtLookUp {
...
+        var decodedResponse = jsonDecode(verbResult);
+        throw AtExceptionUtils.get(
+            decodedResponse['errorCode'], decodedResponse['errorDescription'])
+          ..stack(Triple(
+              Intent.remoteVerbExecution,
+              ExceptionScenario.remoteVerbExecutionFailed,
+              decodedResponse['errorDescription']));
+        // If JSON decoding fails, fall back to older version
+        // Parse the error response in throw AtLookupException.
+      } on FormatException {
```
Comment (gkc) [inline at_lookup/lib/src/at_lookup_impl.dart]: I don't know how expensive it is to try to decode and then throw a FormatException. Could perhaps reduce the likelihood of that happening by checking first of all if the verbResult starts with a '{'? (and only trying to jsonDecode if it does indeed look like json)

**Review by gkc — CHANGES_REQUESTED** [review]

File: at_lookup/lib/src/at_lookup_impl.dart (hunk 2 above)
Comment (sitaram-kalluri) [reply at_lookup/lib/src/at_lookup_impl.dart]: Added IF condition to verify if result startsWith '{' and endsWith '}' to check if response is JSON encoded string.

File: at_lookup/lib/src/at_lookup_impl.dart [hunk 3]
```diff
@@ -267,7 +267,20 @@ class AtLookupImpl implements AtLookUp {
...
     // If response starts with error:, throw AtLookupException.
     if (_isError(verbResult)) {
-      verbResult = verbResult.replaceAll('error:', '');
+      verbResult = verbResult.replaceFirst('error:', '');
+      // If condition to verify if verbResult is a JSON Encoded String.
+      if (verbResult.startsWith('{') && verbResult.endsWith('}')) {
+        try {
+          var decodedResponse = jsonDecode(verbResult);
+          throw AtExceptionUtils.get(decodedResponse['errorCode'],
+              decodedResponse['errorDescription']);
```
Comment (gkc) [inline at_lookup/lib/src/at_lookup_impl.dart:276]: This method (get(errorCode, errorDescription) isn't in the current AtExceptionUtils in at_tools ... can you tell me which PR it is in?

Comment (sitaram-kalluri) [reply at_lookup/lib/src/at_lookup_impl.dart:276]: It is in at_commons. Following is the PR: https://github.com/atsign-foundation/at_tools/pull/186/files

**Review by gkc — APPROVED** [review]

**Comment by sitaram-kalluri** [comment]
Parsing the error responses are deferred to next releases. Hence closing the PR.

**Pull request closed without merging**